arangod/RestServer/TtlFeature.cpp

```diff
@@ -320,66 +341,160 @@ class TtlThread final : public ServerThread<ArangodServer> {
+                transaction::StandaloneContext::Create(*vocbase),
+                aql::QueryString(::lookupQuery), bindVars, options);
+            query->collections().add(collection->name(), AccessMode::Type::READ,
+                                     aql::Collection::Hint::Shard);
+            aql::QueryResult queryResult = query->executeSync();
+
+            if (queryResult.result.fail()) {
+              // we can probably live with an error here...
+              // the thread will try to remove the documents again on next
+              // iteration
+              if (!queryResult.result.is(
+                      TRI_ERROR_QUERY_FORCED_INDEX_HINT_UNUSABLE) &&
+                  !queryResult.result.is(TRI_ERROR_ARANGO_DATABASE_NOT_FOUND) &&
+                  !queryResult.result.is(
+                      TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND)) {
+                LOG_TOPIC("08300", WARN, Logger::TTL)
+                    << "unexpected error during TTL document removal for "
+                       "collection '"
+                    << collection->name()
+                    << "': " << queryResult.result.errorMessage();
+              }
+              break;
+            }
+
+            // finish query so that it does not overlap with following remove
+            // operation
+            query.reset();
+
+            if (queryResult.data == nullptr ||
+                !queryResult.data->slice().isArray()) {
+              break;
+            }
+
+            VPackSlice docsToRemove = queryResult.data->slice();
+            size_t found = docsToRemove.length();
+            if (found == 0) {
+              break;
+            }
+
+            TRI_ASSERT(limitLeft >= found);
+            limitLeft -= found;
+            if (found < maxResultsPerQuery) {
+              // no more documents to remove
+              leftForCurrentCollection = 0;
+            } else {
+              TRI_ASSERT(leftForCurrentCollection >= found);
+              leftForCurrentCollection -= found;
             }
-          } else {
-            auto extra = queryResult.extra;
-            if (extra != nullptr) {
-              VPackSlice v = extra->slice().get("stats");
-              if (v.isObject()) {
-                v = v.get("writesExecuted");
-                if (v.isNumber()) {
-                  uint64_t removed = v.getNumericValue<uint64_t>();
-                  stats.documentsRemoved += removed;
-                  if (removed > 0) {
-                    LOG_TOPIC("2455e", DEBUG, Logger::TTL)
-                        << "TTL thread removed " << removed
-                        << " documents for collection '" << collection->name()
-                        << "'";
-                    if (limitLeft >= removed) {
-                      limitLeft -= removed;
-                    } else {
-                      limitLeft = 0;
-                    }
-                  }
+
+            auto trx = std::make_unique<SingleCollectionTransaction>(
```

Comment:
For a shard belonging to a smart edge collection, we must proceed as follows:
 1. If it is a `_local` shard: to what we do here normally (local removal with replication).
 2. If it is a `_from` shard: send removal as batch to the corresponding global edge collection on some coordinator.
 3. If it is a `_to` shard: skip it.